include/psp2/usbd.h

```diff
@@ -7,89 +7,505 @@
-	SCE_USBD_ERROR_FATAL         = 0x802400FF
+    SCE_USBD_ERROR_NOT_INITIALIZED     = 0x80240001,
+    SCE_USBD_ERROR_ALREADY_INITIALIZED = 0x80240002,
+    SCE_USBD_ERROR_INVALID_PARAM       = 0x80240003,
+
+    SCE_USBD_ERROR_PIPE_NOT_FOUND      = 0x80240004,
+
+    SCE_USBD_ERROR_NO_MEMORY           = 0x80240005,
+    SCE_USBD_ERROR_DEVICE_NOT_FOUND    = 0x80240006,
+
+    SCE_USBD_ERROR_80240007            = 0x80240007, //
+    SCE_USBD_ERROR_80240009            = 0x80240009, //
+    SCE_USBD_ERROR_8024000A            = 0x8024000A, //
+
+    SCE_USBD_ERROR_FATAL               = 0x802400FF
 } SceUsbdErrorCode;
 
 typedef struct SceUsbdDeviceInfo {
-	unsigned int unk0;
-	unsigned int device_id;
-	unsigned int unk2;
-} SceUsbdDeviceInfo; /* size = 0xC */
-
-typedef struct SceUsbdDeviceAddress {
-	unsigned int unk0;
-	unsigned short unk1;
-} SceUsbdDeviceAddress; /* size = 0x6 */
+    unsigned int port;
+    unsigned int device_num;
+    unsigned int unk3; // handled? 0, 1, 2
+} SceUsbdDeviceInfo;
 
 typedef struct SceUsbdTransferData {
-	unsigned int pipe;
-	const void *buff1;
-	unsigned int size1;
-	void *buff2;
-	unsigned int size2;
+    unsigned int pipe; // 0x00
+    const void *data; // 0x04
+    unsigned int data_size;  // 0x08
+    void *transferred; //  // 0x0C // ptr to 8 bytes?
+    unsigned int timeout;  // 0x10 // size of above ptr? only 8 allowed?
 } SceUsbdTransferData; /* size = 0x14 */
 
 typedef struct SceUsbdReceiveEvent {
-	unsigned int unk0;
-	unsigned int unk1;
-	unsigned int unk2;
-	unsigned int unk3;
-	unsigned int unk4;
-	unsigned int unk5;
-	unsigned int transfer_id;
+    unsigned int unk0; // 0
+    unsigned int unk1; // next ptr?
+    unsigned int unk2; // != 8, set to 1, 2, 4? // transfer flags? type?
+    unsigned int unk3; // ??
+    unsigned int unk4; // ??
+    unsigned int unk5; // ??
+    unsigned int transfer_id; // ??
 } SceUsbdReceiveEvent; /* size = 0x1C */
 
+typedef struct SceUsbdDeviceAddress {
+        unsigned int unk0;
+        unsigned short unk1;
+} SceUsbdDeviceAddress; /* size = 0x6 */
+
+#define USB_DESCRIPTOR_DEVICE                   0x01    // bDescriptorType for a Device Descriptor.
+#define USB_DESCRIPTOR_CONFIGURATION            0x02    // bDescriptorType for a Configuration Descriptor.
+#define USB_DESCRIPTOR_STRING                   0x03    // bDescriptorType for a String Descriptor.
+#define USB_DESCRIPTOR_INTERFACE                0x04    // bDescriptorType for an Interface Descriptor.
+#define USB_DESCRIPTOR_ENDPOINT                 0x05    // bDescriptorType for an Endpoint Descriptor.
+
+#define USB_DESCRIPTOR_DEVICE_QUALIFIER         0x06    // bDescriptorType for a Device Qualifier.
+#define USB_DESCRIPTOR_OTHER_SPEED              0x07    // bDescriptorType for a Other Speed Configuration.
+#define USB_DESCRIPTOR_INTERFACE_POWER          0x08    // bDescriptorType for Interface Power.
+#define USB_DESCRIPTOR_OTG                      0x09    // bDescriptorType for an OTG Descriptor.
```

Comment:
can we define enum for these?

Reply:
No

Reply:
can you explain a reason?

Reply:
enums are good for type-safety of func params, for example. but that isn't the case here. those are just "helpers" for human-readable values. defining them as enums would require either changing func and struct param types (which i'm against, since that would break struct compatibility between implementations on other systems) or would require type-casting each time they're used.
same goes for classes below.